app/services/EtmpRegimeService.scala

```diff
@@ -16,33 +16,140 @@
   }
 
-  def checkETMPApi(safeId: String, regime: String)(implicit ec: ExecutionContext, hc: HeaderCarrier): Future[Option[String]] = {
-    if (AWRSFeatureSwitches.regimeCheck().enabled) {
-      getRegimeRefNumber(safeId, regime) recover {
-        case _ => None
-      }
-    } else {
+  def handleDuplicateSubscription(etmpRegistrationDetails: EtmpRegistrationDetails,
+                                  businessCustomerDetails: BusinessCustomerDetails)
+                                 (implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[EtmpRegistrationDetails]] = {
+    authorised(User).retrieve(Retrievals.affinityGroup){ affGroup =>
+      Future(getEtmpRegistrationDetails(affGroup, businessCustomerDetails, etmpRegistrationDetails))
+    } recover {
+      case _ =>
+        None
+    }
+  }
+
+  val AWRS_SERVICE_NAME = "HMRC-AWRS-ORG"
+
+  private def createVerifiers(safeId: String, utr: Option[String], businessType: String, postcode: String) = {
+    val utrTuple = businessType match {
+      case "SOP" => "SAUTR" -> utr.getOrElse("")
+      case _ => "CTUTR" -> utr.getOrElse("")
+    }
+    val verifierTuples = Seq(
+      "Postcode" -> postcode,
+      "SAFEID" -> safeId
+    ) :+ utrTuple
+
+    EnrolmentVerifiers(verifierTuples: _*)
+  }
+
+  def upsertEacdEnrolment(safeId: String,
+                          utr: Option[String],
+                          businessType: String,
+                          postcode: String,
+                          awrsRefNumber: String)(implicit hc: HeaderCarrier): Future[HttpResponse] = {
+    val enrolmentKey = s"$AWRS_SERVICE_NAME~AWRSRefNumber~$awrsRefNumber"
+    val enrolmentVerifiers = createVerifiers(safeId, utr, businessType, postcode)
+    enrolmentStoreConnector.upsertEnrolment(enrolmentKey, enrolmentVerifiers)
+  }
+
+  def checkETMPApi(safeId: String, businessCustomerDetails: BusinessCustomerDetails, businessRegistrationDetails: BusinessRegistrationDetails)
+                  (implicit ec: ExecutionContext, hc: HeaderCarrier): Future[Option[EtmpRegistrationDetails]] = {
+    val businessType = businessRegistrationDetails.legalEntity.getOrElse("")
+    val postcode = businessCustomerDetails.businessAddress.postcode.getOrElse("").replaceAll("\\s+", "")
+
+    if (!AWRSFeatureSwitches.regimeCheck().enabled) {
       Future.successful(None)
+    } else {
+      getEtmpBusinessDetails(safeId) flatMap {
+        case Some(etmpRegDetails) =>
+          handleDuplicateSubscription(etmpRegDetails, businessCustomerDetails) flatMap { _ =>
+            upsertEacdEnrolment(
+              safeId,
+              businessRegistrationDetails.utr,
+              businessType,
+              postcode,
+              etmpRegDetails.regimeRefNumber
+            ) map { response =>
+              response.status match {
+                case NO_CONTENT => Some(etmpRegDetails)
+                case _ =>
+                  Logger.warn("[EtmpRegimeService][checkETMPApi] Failed to upsert to EACD")
```

Comment:
AMBER - Adding the returned status to this log could be helpful to us in the future.